### ISSUE TITLE
False error reporting in properties files for enum types

### ISSUE THREAD

**Issue by T3rm1** [spring-projects/sts4#215]
`dwh.proxy.client.meta-data.type=web_application`

    public enum ApplicationType {
		WEB_APPLICATION,
		WEB_SERVICE,
		NONE
	}

Error is `'web_application' is an unknown dwhproxy.client.configuration.DwhProxyClientProperties$ApplicationType[NONE, WEB_APPLICATION, WEB_SERVICE]'. Valid values are: [NONE, WEB_APPLICATION, WEB_SERVICE, none, web-application, web-service]`

However this is no error. `web_application` with underscore works.

**Comment by kdvolder** [comment]
This is a bit of a conundrum unfortunately. 

The rules spring boot uses for these 'relaxed names' are, unfortunately a bit nebulously defined. To make matters worse they tend to change slightly over time when implementation of how they are handled internally is changed. 

To remain practical, STS has supported only the two variants that seemed most commonly used. 1) use the upper-case + underscores name that is exactly as it is typically used to name the enum values in Java. and 2) use the 'canonical' hyphenated, lowercased version of the name..

We could certainly add another 'variant' of these relaxed names. But I'm a bit reluctant to do that. How important do you think it is? E.g. why can't you just simply use one of the already supported name variants?

**Comment by T3rm1** [comment]
I can use one or the other variant. That's not a problem. However I thought the validation should be adjusted to reflect the internals of parsing class in Spring.
I don't think this will be changed so fast in the future.

**Comment by kdvolder** [comment]
> However I thought the validation should be adjusted to reflect the internals of parsing class in Spring.

In principle I agree with you. However I have studied the implementation a few times as well as read the documentation. The implementation is quite complex and the 'variants' that it generates can't be so easily characterised with simple rules. It would complex to try and replicate exactly, and it is unclear how much benefit this really brings. 

> I don't think this will be changed so fast in the future.

Maybe not. We don't really know. I do know however that it has changed more than once in the past already. And since the docs are really not pinning all the variants down exactly, it can be argued that the behavior is actually not completely defined and some of the variants that the parser accepts are more of an accident of its implementation than something intended to preserved forever into the future.

My take on it is that you would probably do well to use the 'canonical' hyphenated form instead of a mix of the uppercase and hyphenated forms. I am pretty sure the two variants the editor currently accepts will allways work for both older and newer versions of boot. I wouldn't be so sure about the 'mixed' variant.

So, if you agree that it isn't too honerous on a user to just use one of those two variants, which probably a good idea anyway, then I will just close this as "Won't fix".

If you can find some wording in the docs that clearly indicate the mixed variant (lower case + underscores) should also work than I will re-consider. 

In other words, the test we want to apply is 'do the docs say that this should work?'. if yes, then we should probably support it. If no, then I consider it more of 'accident' that it works, and the editor probably does well to discourage you from depending on this.
